File: src/2025-03-04-12-20-21--deepseek-vr-0304/python/exercises/practice/affine-cipher/affine_cipher.py
import math

def encode(plain_text, a, b):
    if math.gcd(a, 26) != 1:
        raise ValueError("a and m must be coprime.")
    filtered = []
    for c in plain_text:
        if c.isalnum():
            filtered.append(c.lower() if c.isalpha() else c)
    encrypted = []
    for char in filtered:
        if char.isalpha():
            i = ord(char) - ord('a')
            encrypted_char = chr((a * i + b) % 26 + ord('a'))
            encrypted.append(encrypted_char)
        else:
            encrypted.append(char)
    encrypted_str = ''.join(encrypted)
    groups = [encrypted_str[i:i+5] for i in range(0, len(encrypted_str), 5)]
    return ' '.join(groups)

def decode(ciphered_text, a, b):
    if math.gcd(a, 26) != 1:
        raise ValueError("a and m must be coprime.")
    a_inv = pow(a, -1, 26)
    ciphered = ciphered_text.replace(' ', '')
    decrypted = []
    for char in ciphered:
        if char.isalpha():
            y = ord(char) - ord('a')
            decrypted_char = chr((a_inv * (y - b)) % 26 + ord('a'))
            decrypted.append(decrypted_char)
        else:
            decrypted.append(char)
    return ''.join(decrypted)

File: src/2025-03-04-12-20-21--deepseek-vr-0304/java/exercises/practice/affine-cipher/src/main/java/AffineCipher.java
public class AffineCipher {
    
    private int gcd(int a, int b) {
        a = Math.abs(a);
        b = Math.abs(b);
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    private int modInverse(int a, int m) {
        a = Math.floorMod(a, m);
        for (int x = 1; x < m; x++) {
            if (Math.floorMod(a * x, m) == 1) {
                return x;
            }
        }
        throw new IllegalArgumentException("No modular inverse exists.");
    }

    public String encode(String text, int coefficient1, int coefficient2) {
        int a = coefficient1;
        int b = coefficient2;
        if (gcd(a, 26) != 1) {
            throw new IllegalArgumentException("Coefficient 'a' must be coprime with 26.");
        }

        StringBuilder processed = new StringBuilder();
        for (char c : text.toCharArray()) {
            if (Character.isLetter(c)) {
                processed.append(Character.toLowerCase(c));
            } else if (Character.isDigit(c)) {
                processed.append(c);
            }
        }
        String filtered = processed.toString();

        StringBuilder result = new StringBuilder();
        for (char c : filtered.toCharArray()) {
            if (Character.isLetter(c)) {
                int x = c - 'a';
                int encrypted = Math.floorMod(a * x + b, 26);
                result.append((char) (encrypted + 'a'));
            } else if (Character.isDigit(c)) {
                result.append(c);
            }
        }

        // Group into 5-letter chunks
        StringBuilder grouped = new StringBuilder();
        int count = 0;
        for (char c : result.toString().toCharArray()) {
            grouped.append(c);
            count++;
            if (count == 5) {
                grouped.append(' ');
                count = 0;
            }
        }
        return grouped.toString().trim();
    }

    public String decode(String text, int coefficient1, int coefficient2) {
        int a = coefficient1;
        int b = coefficient2;
        if (gcd(a, 26) != 1) {
            throw new IllegalArgumentException("Coefficient 'a' must be coprime with 26.");
        }

        int aInv = modInverse(a, 26);

        StringBuilder filtered = new StringBuilder();
        for (char c : text.toCharArray()) {
            if (Character.isLetterOrDigit(c)) {
                filtered.append(Character.toLowerCase(c));
            }
        }
        String input = filtered.toString();

        StringBuilder result = new StringBuilder();
        for (char c : input.toCharArray()) {
            if (Character.isLetter(c)) {
                int y = c - 'a';
                int decrypted = Math.floorMod(aInv * (y - b), 26);
                result.append((char) (decrypted + 'a'));
            } else if (Character.isDigit(c)) {
                result.append(c);
            }
        }

        return result.toString();
    }
}
